examples/mnist_example.py

```diff
@@ -0,0 +1,259 @@
+    parser.add_argument('--out_dir',
+                        type=str,
+                        help='Path where the trained model will be saved (if not None).')
+    args = parser.parse_args()
+
+    torch.manual_seed(args.seed)
+    accelerator = Accelerator()
+    tqdm = partial(original_tqdm, disable=not accelerator.is_main_process, position=0)
+
+    # TRAIN AND TEST DATASETS/DATALOADERS
+    train_transforms = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize((0.1307,), (0.3081,)),
+    ])
+
+    test_transforms = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize((0.1307,), (0.3081,)),
+    ])
+
+    # Download MNIST only on main process
+    with accelerator.main_process_first():
+
+        train_dataset = datasets.MNIST(os.environ['DSDIR'],
+                                       train=True,
+                                       download=True,
+                                       transform=train_transforms)
+
+        test_dataset = datasets.MNIST(os.environ['DSDIR'],
+                                      train=False,
+                                      download=True,
+                                      transform=test_transforms)
+
+    train_dataloader = DataLoader(dataset=train_dataset,
+                                  batch_size=args.per_device_train_batch_size,
+                                  shuffle=True,
+                                  pin_memory=True)
+
+    test_dataloader = DataLoader(dataset=test_dataset,
+                                 batch_size=args.per_device_eval_batch_size,
+                                 shuffle=True,
+                                 pin_memory=True)
+
+    model = Net().to(accelerator.device)
+    optimizer = optim.Adadelta(model.parameters(), lr=args.lr)
+    scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
+    model, optimizer, train_dataloader, test_dataloader = accelerator.prepare(model, optimizer, train_dataloader, test_dataloader)
+
+    def evaluate(_model: nn.Module,
+                 _device: Union[torch.device, str],
+                 _test_loader: DataLoader,
+                 _epoch: int):
+        _model.eval()
+
+        test_accuracy = Accuracy()
+
+        with torch.no_grad():
+            for data, target in tqdm(_test_loader, desc=f'Eval (epoch {_epoch:03d})'):
+
+                data, target = data.to(_device, non_blocking=True), target.to(_device, non_blocking=True)
+                output = _model(data)
+                preds = output.argmax(dim=1, keepdim=True)
+
+                # Compute accuracy:
+                # We first gather predictions & labels from all processes
+                bs = _test_loader.batch_size
+                y_pred = accelerator.gather(preds).detach().cpu()[:bs]
+                y_true = accelerator.gather(target).detach().cpu()[:bs]
+                test_accuracy.update((y_pred, y_true))
+
+        test_acc = test_accuracy.compute()
```

Comment:
I'd prefer to see all the logic to compute the accuracy here. We don't need a generic class for this in this particular example, and it would make it easier to read I believe.